Added beats as a transition unit

karanimwenda opened the pull request:
I have added beats as a unit for specifying autodj's transition duration. It is wishlisted [here](https://bugs.launchpad.net/mixxx/+bug/1384873?comments=all).

karanimwenda committed:
dlgprefautodjdlg.ui : Added transition unit option

Added a combobox for specifying the units to use
when specifying how long a transition should last

karanimwenda committed:
DlgPrefAutoDj: Now storing the set transition unit

Added the logic for setting the preferred transition unit
to use: i.e. beats or seconds

karanimwenda committed:
AutoDJProcessor: Getting the set transition unit

Getting the transition unit set in the preferences.

karanimwenda committed:
DlgAutoDJ: updating labelTransitionAppendix's text

Retrieving the transition unit set in the preferences then
updating the value accordingly

karanimwenda committed:
AutoDJProcessor: Calculating fadeTime when both syncs enabled

Simply converting the beats to seconds like:
timeInSecs = (nbrBeats * 60) / bpm

karanimwenda committed:
AutoDJProcessor: Using getFadeTime fn in calculateTransition

The transition time can be expressed in beats or seconds.
The function getFadeTime ensures calculateTransition gets
the transition time in seconds.

karanimwenda committed:
dlgprefautodjdlg.ui: Added a combobox to select which deck's bpm to use

This information will be key when sync isn't enabled on both decks
as the BPMs won't be equal.

karanimwenda committed:
DlgPrefAutoDj: Added a combobox that specifies which deck's BPM to use

This combobox is only enabled when the user sets 'beats' as the
transition unit.

karanimwenda committed:
AutoDJProcessor: Using the BPM from a user-preferred deck

Using the BPM from the user-preferred deck to transition
when both decks do not have sync enabled.

karanimwenda committed:
AutoDJProcessor: Returning a signed value from getFadeTime

This will allow negative transition times.

JosepMaJAZ commented on src/library/autodj/autodjprocessor.cpp:
```diff
@@ -1659,3 +1678,83 @@ bool AutoDJProcessor::nextTrackLoaded() {
+
+double AutoDJProcessor::getFadeTime() {
+    // TODO(c3n7): Is there a better place for this? Find out.
+    // Copied from syncable.h
```
If it is already on a file that you can include, why do you copy its content?

c3n7 replied:
[Done](https://github.com/mixxxdj/mixxx/pull/2746/commits/c2f9ad507c0407d0d8ab45b0403873cafd6fba8a). Included `syncable.h`.

JosepMaJAZ commented on src/library/autodj/autodjprocessor.cpp:
```diff
@@ -1659,3 +1678,83 @@ bool AutoDJProcessor::nextTrackLoaded() {
+
+    if (m_transitionUnit == TransitionUnit::BEATS) {
+        if ((leftSyncMode == SYNC_FOLLOWER && rightSyncMode == SYNC_MASTER_SOFT) ||
+                (leftSyncMode == SYNC_MASTER_SOFT && rightSyncMode == SYNC_FOLLOWER)) {
```
The master sync feature is being improved in another PR, this condition might ignore some cases (when there is an explicit master), even though it is not too problematic, because the else case would still work.

c3n7 replied:
What about simply checking if both decks have the same BPM? I think that will achieve the same effect too while avoiding any conflicts.

c3n7 replied:
[Done](https://github.com/mixxxdj/mixxx/pull/2746/commits/938dfe87e839ad6abe76ec26f6fb5d4120d32186). Now the condition is simply checking whether both tracks' BPM are equal.

JosepMaJAZ replied:
Yes, that's a better way to test this

JosepMaJAZ commented on src/library/autodj/autodjprocessor.cpp:
```diff
@@ -1659,3 +1678,83 @@ bool AutoDJProcessor::nextTrackLoaded() {
+                (leftSyncMode == SYNC_MASTER_SOFT && rightSyncMode == SYNC_FOLLOWER)) {
+            // Both the left and right bpm should be equal if this point is reached
+            VERIFY_OR_DEBUG_ASSERT(leftBPM == rightBPM) {
+                qDebug() << tr("leftBPM: %1   - rightBPM %2").arg(leftBPM).arg(rightBPM);
```
this should probably be a qWarning()

c3n7 replied:
[Done. ](https://github.com/mixxxdj/mixxx/pull/2746/commits/f025f8a624e3c44a12ef8dc71e585df72eee03a5)Now using `qWarning()`.

JosepMaJAZ commented on src/library/autodj/dlgautodj.cpp:
```diff
@@ -283,6 +291,16 @@ void DlgAutoDJ::toggleAutoDJButton(bool enable) {
+
+    // Update the label to display the configured transition unit
+    // just in case it was changed in the preferences dialog
+    QString transitionUnit;
```
Please, do not duplicate this code, make a method with its content.
This is specially important since it includes a translation, and you would be forcing to translate it twice.

c3n7 replied:
[Done](https://github.com/mixxxdj/mixxx/pull/2746/commits/c2f9ad507c0407d0d8ab45b0403873cafd6fba8a). Moved duplicate code to a function.

JosepMaJAZ commented on src/preferences/dialog/dlgprefautodj.cpp:
```diff
@@ -184,3 +247,52 @@ void DlgPrefAutoDJ::slotEnableAutoDJRandomQueue(int a_iValue) {
+    switch (a_iValue) {
+    case 1:
+        // To-deck's BPM
+        m_pConfig->set(ConfigKey("[Auto DJ]", "BPMToUseBuff"),
```
Since you are using an int here, you could use the setValue instead.

c3n7 replied:
[Done](https://github.com/mixxxdj/mixxx/pull/2746/commits/61fcc3f4bd44a407707731ab972bc0fd005ab0f1). Using setValue instead of set.

JosepMaJAZ reviewed (changes requested):
Nice addition. When using sync, this was something that felt missing for so long.

I added some comments. I will try to test it later today.

karanimwenda committed:
AutoDJProcessor: Using qWarning to display a message in an assert

karanimwenda committed:
DlgPrefAutoDJ: Switched to using setValue to store an integer value

karanimwenda committed:
AutoDJProcessor: Included syncable.h to get access to the SyncMode enum

JosepMaJAZ reviewed (commented):
After testing it, I believe that the descriptions are a bit misleading.

Maybe it is just me, but "From deck's bpm", "To deck's bpm", "which deck's bpm auto dj should use"... 
All these suggested that it was really synchronizing the BPM (similarly to a fake-controller script that we have on the forums which adapts the BPM while fading from one track to another).
https://mixxx.org/forums/viewtopic.php?f=7&t=8318&hilit=auto+dj+bpm+script#p34531

What I've been able to test, and what the code actually does is determine which BPM value will it use to calculate the fading duration 
i.e.:  determine if "8 beats" means "before track's end" or "since next track's start".

Of course, both are the same if both tracks are synchronized and have the same BPM.

I believe the correct option would be:
If tracks don't have the same BPM, calculate beats:
- from track's end
- from next track's start

I can't see the usefulness of specifying a fixed deck. It would cause that on one track obeys the track end and on the next track obeys the next track start

Oh.. and if you what to implement also an option to adapt from one BPM to the next... :) (But that would complicate the calculation of the transition time) You can check the script for information.

karanimwenda committed:
DlgAutoDJ: Moved duplicate code to a function

c3n7 commented:
You are right. My description is a bit misleading. I am pushing a commit right about now. I will definitely work on adapting from one BPM to the next. The script you have linked will definitely be helpful. Thank you. While playing with Auto DJ I felt like that feature would be very convenient.

karanimwenda committed:
Using more descriptive transition enums

DlgPrefAutoDJ, AutoDJProcessor:
Used enums which describe better how transition periods will
be calculated if the tracks' BPM aren't equal

karanimwenda committed:
AutoDJProcessor: Now simply checking if both tracks' BPM are equal.

karanimwenda committed:
AutoDJProcessor: Removed unused variables

JosepMaJAZ commented:
Overall it looks ok now, but I still have one issue with the "secs/beats" label not updating when I expect it to.
As I saw, it can only refresh on two situations:  on application startup (since it's on the constructor of AutoDJ), or when toggling autodj mode.

I find this confusing, since after changing the setting, I would discover this just by chance. 
One solution could be to add a signal to call from dlgprefautodj to inform autoDJ (both the processor and the dialog) that the setting has changed, but I'm not sure how easy this could be (since you need to connect the slots in order to listen to that signal).

c3n7 commented:
True, currently the label is only updated in DlgAutoDj's constructor and when Auto DJ is enabled. I agree a signal/slot approach would be better. Once I know how to get that working I will push a commit.

karanimwenda committed:
DlgAutoDJ.ui: Added UI to switch between beats and seconds

karanimwenda committed:
AutoDjProcessor, DlgAutoDjProcessor: copied corresponding code from the preferences

karanimwenda committed:
AutoDJProcessor: Recalculating transitions on transition-unit change

karanimwenda committed:
DlgPrefAutoDJ: Removed transition-unit preference-setting ui

The ui is now in the main window

katsar0v commented:
Works really well. The dropdowns and the fields need to be adjusted, to be in the design of the skin.

LateNight:

![image](https://user-images.githubusercontent.com/7236157/86515373-b269eb80-be18-11ea-9009-2cca34e47ee8.png)

Deere:

![image](https://user-images.githubusercontent.com/7236157/86515376-bb5abd00-be18-11ea-8b9a-ad7b1582acbd.png)

karanimwenda committed:
Deere skin: Styling the added comboboxes

`ComboBoxAutoDjTransitionUnit` and `ComboBoxAutoDjCalcTransitionPeriod` now
share the same styling with `fadeModeCombobox`

karanimwenda committed:
Merge branch 'master' of https://github.com/mixxxdj/mixxx into bars-transition-unit

karanimwenda committed:
LateNight skin: Styled added comboboxes

ComboBoxAutoDjTransitionUnit and ComboBoxAutoDjCalcTransitionPeriod now
share the same styling with fadeModeCombobox.

Be-ing commented:
I haven't looked at the code yet, but I just tested and my first impression is pretty good. The UI could use work though. The new QComboBox for selecting the transition time units is redundant with the unit label text next to the spinbox. How about moving the QComboBox to the right of the spinbox and getting rid of the redundant text?

Also, I expected the QDoubleSpinbox to halve and double like it does for the loop & beatjump sizes instead of incrementing/decrementing by 1. You could subclass WBeatSpinBox to make a new widget that switches between doubling/halving for beats and +1/-1 for seconds.

Be-ing commented on src/library/autodj/dlgautodj.cpp:
```diff
@@ -145,6 +145,11 @@ DlgAutoDJ::DlgAutoDJ(
+    QString transitionUnitToolTip = tr(
+            "Sets the units to use when specifying how long a transition should last.\n");
+    QString calcTransitionPeriodToolTip = tr(
+            "If the tracks do not have the same BPM, this specifies\n"
```
```suggestion
            "When using beats for the transition time, if the tracks do not have the same BPM, this specifies\n"
```

Be-ing commented on src/library/autodj/autodjprocessor.h:
```diff
@@ -168,6 +168,16 @@ class AutoDJProcessor : public QObject {
     };
 
+    enum class TransitionUnit {
+        SECONDS,
+        BEATS
```
```suggestion
        Seconds,
        Beats
```

Be-ing commented on src/library/autodj/dlgautodj.ui:
```diff
@@ -115,16 +115,26 @@
        </property>
+      </widget>
+     </item>
+     <item>
+      <widget class="QComboBox" name="ComboBoxAutoDjCalcTransitionPeriod"/>
+     </item>
```
How about instead of adding this QComboBox we make separate options for the transition unit combobox:
  * Seconds
  * Beats (outgoing track)
  * Beats (incoming track)

c3n7 replied:
Makes sense, I'll push an update with those changes.

Be-ing commented on src/library/autodj/autodjprocessor.cpp:
```diff
@@ -1659,3 +1681,67 @@ bool AutoDJProcessor::nextTrackLoaded() {
+
+    if (m_transitionUnit == TransitionUnit::BEATS) {
+        if (leftBPM == rightBPM) {
+            fadeTime = (transitionTime * 60) / leftBPM;
```
This assumes that the BPM is constant. @hacksdump and @crisclacerda are working hard to make Mixxx work better with tracks that have changing tempos. Instead of getting the BPM from a ControlProxy, AutoDJProcessor should get the Beats object from the Track object when a Track is loaded and call Beats' method for finding the beats.

c3n7 replied:
What if the user changes the BPM using a Deck's RateSlider, will the Beats object from the Track object contain the updated value?

Be-ing replied:
That's a great question. I don't think so. So you may need to multiple the positions from Beats by the `rate_ratio` ControlProxy.

c3n7 replied:
Thanks, I'll work on that.

Be-ing replied:
That would still assume the tempo stays constant over the transition. :/ For AutoDJ, I think that's a good enough hack because it is *Auto*DJ, so the user probably isn't adjusting the tempo as the transition happens. If we ever make AutoDJ automatically sync and fade tempos that will be significantly more complicated.

Be-ing replied:
You could connect to `valueChanged` signals of `rate_ratio` and recalculate the transition times.

Be-ing replied:
Oh, we already do this in AutoDJProcessor::playerRateChanged. :smile:

Be-ing replied:
> Instead of getting the BPM from a ControlProxy, AutoDJProcessor should get the Beats object from the Track object when a Track is loaded and call Beats' method for finding the beats.

Ping. This still needs to be implemented.

c3n7 replied:
I'll work on it and hopefully get it working by the end of this day.

c3n7 replied:
I hope [commit e02deecce](https://github.com/mixxxdj/mixxx/pull/2746/commits/e02deecce247f59c9f22df9111c7aa59103e5e86) does this well.

Holzhaus commented:
Do we have end silence detection? Otherwise use beats as transition unit might be problematic In some cases.

Be-ing commented:
> Do we have end silence detection?

Yes, but simply counting beats from the beginning or end of the track does not take that into account.

> use beats as transition unit might be problematic In some cases

Of course, it's AutoDJ. Any automatic mixing will be problematic in some cases without manual setup.

Holzhaus commented:
How about counting beats from the intro start/outro end cues instead? These are autodetected by default and I think this would make this a lot more useful. If these Cues are not set, we could fall back to counting beats from the actual start/end.

Be-ing commented:
If you want to use the intro and outro cues for AutoDJ, then use one of the intro/outro modes. I don't think this should have anything to do with those.

Holzhaus commented:
It was my impression that this feature is useful for  people who don't want to set intro/outro points manually.

Holzhaus commented:
Another idea: can we store the silence end/start positions in addition to the cue in/out markers and use them here? That way this feature is independent of the cue in/out markers while still giving expected behavior.

Holzhaus commented:
> Another idea: can we store the silence end/start positions in addition to the cue in/out markers and use them here? That way this feature is independent of the cue in/out markers while still giving expected behavior. 

I still think this would be useful, but can be taken care of later on.

Holzhaus commented on src/library/autodj/autodjprocessor.cpp:
```diff
@@ -1393,13 +1410,14 @@ void AutoDJProcessor::useFixedFadeTime(
         double fadeEndSecond,
         double toDeckStartSecond) {
-    if (m_transitionTime > 0.0) {
+    double fadeTime = getFadeTime();
```
```suggestion
    const double fadeTime = getFadeTime();
```

Holzhaus commented on src/library/autodj/autodjprocessor.cpp:
```diff
@@ -1659,3 +1681,67 @@ bool AutoDJProcessor::nextTrackLoaded() {
 }
+
+double AutoDJProcessor::getFadeTime() {
+    double leftBPM = m_pLeftBPM->get();
+    double rightBPM = m_pRightBPM->get();
```
```suggestion
    const double leftBPM = m_pLeftBPM->get();
    const double rightBPM = m_pRightBPM->get();
```

Holzhaus commented on src/library/autodj/autodjprocessor.cpp:
```diff
@@ -1659,3 +1681,67 @@ bool AutoDJProcessor::nextTrackLoaded() {
+    double rightBPM = m_pRightBPM->get();
+    double fromBPM, toBPM;
+
+    double transitionTime = m_transitionTime;
```
```suggestion
    const double transitionTime = m_transitionTime;
```

Holzhaus commented on src/library/autodj/autodjprocessor.cpp:
```diff
@@ -263,7 +279,7 @@ void AutoDJProcessor::fadeNow() {
     // If the user presses "Fade now", assume they want to fade *now*, not later.
     // So if the spinbox time is negative, do not insert silence.
-    double spinboxTime = fabs(m_transitionTime);
+    double spinboxTime = fabs(getFadeTime());
```
```suggestion
    const double spinboxTime = fabs(getFadeTime());
```

karanimwenda committed:
AutoDJProcessor: Made vars that don't change often constants

Co-authored-by: Jan Holthuis <holthuis.jan@googlemail.com>

karanimwenda committed:
AutoDJProcessor: Made vars that don't change often constants

Co-authored-by: Jan Holthuis <holthuis.jan@googlemail.com>

karanimwenda committed:
AutoDJProcessor: Made vars that don't change often constants

Co-authored-by: Jan Holthuis <holthuis.jan@googlemail.com>

karanimwenda committed:
AutoDJProcessor: Made vars that don't change often constants

Co-authored-by: Jan Holthuis <holthuis.jan@googlemail.com>

karanimwenda committed:
AutoDjProcessor: Made a tooltip more descriptive

Co-authored-by: Be <be.0@gmx.com>

karanimwenda committed:
AutoDjProcessor: Changed transition unit enum from Upper to Title Case

Co-authored-by: Be <be.0@gmx.com>

karanimwenda committed:
Merge branch 'master' into bars-transition-unit

karanimwenda committed:
DlgAutoDJ and AutoDJProcessor: Updated enum changes

SECONDS and BEATS to Seconds and Beats, respectively

karanimwenda committed:
DlgAutoDj: Set the transition unit in one combobox

karanimwenda committed:
DlgAutoDJ: Fixed Styling Issues

karanimwenda committed:
Deere: Removed trailing whitespace

karanimwenda committed:
AutoDjProcessor & DlgAutoDj: Combined Combobox fns

Moved ComboBoxAutoDjCalcTransitionPeriod's functionality
to ComboBoxAutoDjTransitionUnit

karanimwenda committed:
DlgAutoDJUI: Removed unused combobox

karanimwenda committed:
LateNight: Removed ComboBoxAutoDjCalcTransitionPeriod

karanimwenda committed:
Deere: Removed ComboBoxAutoDjCalcTransitionPeriod

karanimwenda committed:
DlgAutoDJ: Added spacing to tooltip

Be-ing commented:
The GUI is better now. :+1:

karanimwenda committed:
Shade: Added ComboBoxAutoDjTransitionUnit

karanimwenda committed:
Tango: Added ComboBoxAutoDjTransitionUnit

Be-ing commented:
![image](https://user-images.githubusercontent.com/9455094/97191682-1e2fe580-1775-11eb-9806-736b9cfa3f61.png)
The "beats/sec" label to the right of the spinbox is redundant. I think it would be better to move the unit selection combobox to the right of the spinbox and add a small spacer between the mix mode combobox and the spinbox.

karanimwenda committed:
AutoDJProcessor: Getting the BPM Straight from the Track

c3n7 commented:
Makes sense. I'm working on removing the "beats/sec" label.

Be-ing commented on src/library/autodj/autodjprocessor.cpp:
```diff
@@ -1674,6 +1669,18 @@ double AutoDJProcessor::getFadeTime() {
+    TrackPointer leftTrack = pLeftDeck->getLoadedTrack();
+    TrackPointer rightTrack = pRightDeck->getLoadedTrack();
+
+    const double leftTrackBPM = leftTrack->getBpm();
+    const double rightTrackBPM = rightTrack->getBpm();
```
This gets an average BPM for the whole track so it still assumes a constant tempo. AutoDJProcessor should not be concerned with the tempo for this feature; it only needs the length of time that the last beats of the track occupy (multiplied by rateRatio). This information must come from the Beats object.

c3n7 replied:
Interesting. Thanks. I'll work on that.

c3n7 replied:
Hope in commit  [33f83a4](https://github.com/mixxxdj/mixxx/pull/2746/commits/33f83a4ed49b2401a1e6d51599dce568cb2f780d) I'm moving in the right direction?

karanimwenda committed:
DlgAutoDJ: Removed labelTransitionAppendix

karanimwenda committed:
Tango: removed labelTransitionAppendix

karanimwenda committed:
AutoDJProcessor: Calculating fadeTime for FixedFull Transition

karanimwenda committed:
Merge branch 'main' into bars-transition-unit

karanimwenda committed:
Merge branch 'main' of https://github.com/mixxxdj/mixxx into bars-transition-unit

 Conflicts:
	res/skins/Deere/style.qss
        res/skins/LateNight/style_palemoon.qss
        res/skins/Tango/style.qss
        src/library/autodj/autodjprocessor.h
        src/library/autodj/dlgautodj.ui

Be-ing commented:
The build is now failing with unused parameter warnings.
```
 [ 39%] Building CXX object CMakeFiles/mixxx-lib.dir/src/library/autodj/dlgautodj.cpp.o
/home/runner/work/mixxx/mixxx/src/library/autodj/autodjprocessor.cpp: In member function ‘double AutoDJProcessor::getFadeTimeArg(DeckAttributes*, DeckAttributes*, double, double, double)’:
Error: /home/runner/work/mixxx/mixxx/src/library/autodj/autodjprocessor.cpp:1685:16: error: unused parameter ‘fromDeckSecond’ [-Werror=unused-parameter]
 1685 |         double fromDeckSecond,
      |         ~~~~~~~^~~~~~~~~~~~~~
Error: /home/runner/work/mixxx/mixxx/src/library/autodj/autodjprocessor.cpp:1686:16: error: unused parameter ‘fadeEndSecond’ [-Werror=unused-parameter]
 1686 |         double fadeEndSecond,
      |         ~~~~~~~^~~~~~~~~~~~~
Error: /home/runner/work/mixxx/mixxx/src/library/autodj/autodjprocessor.cpp:1687:16: error: unused parameter ‘toDeckStartSecond’ [-Werror=unused-parameter]
 1687 |         double toDeckStartSecond) {
      |         ~~~~~~~^~~~~~~~~~~~~~~~~
cc1plus: all warnings being treated as errors
make[2]: *** [CMakeFiles/mixxx-lib.dir/build.make:2442: CMakeFiles/mixxx-lib.dir/src/library/autodj/autodjprocessor.cpp.o] Error 1
make[2]: *** Waiting for unfinished jobs....
make[1]: *** [CMakeFiles/Makefile2:1374: CMakeFiles/mixxx-lib.dir/all] Error 2
make: *** [Makefile:163: all] Error 2
Error: Process completed with exit code 2.
```

c3n7 commented:
Those variables I'm only using in the debugger to test some stuff, I'll get rid of them once I'm done.  
I can see a lot has been done, I was ~2500 commits behind. All in all I think the strict compiler settings are nice.

uklotzde commented:
> All in all I think the strict compiler settings are nice.

That's the right attitude, Timothy :+1:

ronso0 commented:
@c3n7 Any chance you're still working on this feature?

c3n7 commented:
@ronso0 It's been quite a long while since I did my last commit, a lot has been happening but I'll dedicate at least a day every week till I get it done. Meanwhile, I'm looking at others pull requests so I can familiarize myself again with the codebase.
